File: src/hooks/excel-api/references.ts
import { proxyJson } from '@/hooks/excel-api/http'
import type { createHealthApi } from '@/hooks/excel-api/health'

type HealthApi = ReturnType<typeof createHealthApi>

export function createReferencesApi({ ensureExcelBackend }: HealthApi) {
  const referencesStatus = async () => {
    await ensureExcelBackend({ lenient: true })
    return proxyJson('/api/excel/references?action=status', { method: 'GET' })
  }

  const referencesLoad = async () => {
    await ensureExcelBackend()
    return proxyJson('/api/excel/references?action=load', { method: 'POST' })
  }

  const referencesApply = async () => {
    await ensureExcelBackend()
    return proxyJson('/api/excel/references?action=apply', { method: 'POST' }, 600_000)
  }

  const referencesUpload = async (kind: 'territory' | 'podr' | 'login', file: File) => {
    await ensureExcelBackend()
    const form = new FormData()
    form.append('file', file)
    const res = await fetch(`/api/excel/references/upload?kind=${kind}`, {
      method: 'POST',
      body: form,
    })
    const data = await res.json().catch(() => ({}))
    if (!res.ok) {
      const detail = (data as { detail?: string }).detail
      throw new Error(
        detail
        || (data as { error?: string }).error
        || (res.status === 404
          ? 'Сервис справочников не найден — перезапустите excel-backend (RESTART-EXCEL.bat)'
          : 'Ошибка загрузки'),
      )
    }
    return data
  }

  /** Проверка файлов на диске без Python (если бэкенд ещё не обновлён). */
  const referencesLocalStatus = async () => {
    const res = await fetch('/api/excel/references/local-status')
    const data = await res.json().catch(() => ({}))
    if (!res.ok) {
      throw new Error((data as { error?: string }).error || 'Не удалось прочитать каталог справочников')
    }
    return data as {
      references_dir: string
      files: Record<string, boolean>
      resolved_files?: Record<string, string>
    }
  }

  return {
    referencesStatus,
    referencesLoad,
    referencesApply,
    referencesUpload,
    referencesLocalStatus,
  }
}

File: mini-services/excel-service/routers/references.py
"""Справочники API."""
from __future__ import annotations

import os

from fastapi import APIRouter, HTTPException, UploadFile, File
from pydantic import BaseModel

from deps import run_blocking
import references

router = APIRouter()


class LoginRequest(BaseModel):
    login: str
    password: str


@router.post("/api/auth/login")
async def auth_login(body: LoginRequest):
    try:
        user = await run_blocking(references.verify_user, body.login, body.password)
        if not user:
            raise HTTPException(status_code=401, detail="Неверный логин или пароль")
        return {
            "user": {
                "login": user["login"],
                "role": user["role"],
                "sites": user.get("sites") or [],
            },
        }
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/api/references/status")
async def references_status():
    try:
        return await run_blocking(references.status)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/api/references/load")
async def references_load():
    try:
        return await run_blocking(references.load_from_disk)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/api/references/apply")
async def references_apply():
    try:
        result = await run_blocking(references.apply_all)
        err = (result.get("main_db") or {}).get("error")
        if err:
            raise HTTPException(status_code=400, detail=err)
        return result
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/api/references/upload/{kind}")
async def references_upload(kind: str, file: UploadFile = File(...)):
    kind_map = {
        "territory": references.TERRITORY_FILE,
        "podr": references.PODR_FILE,
        "login": references.LOGIN_FILE,
    }
    target_name = kind_map.get(kind)
    if not target_name:
        raise HTTPException(status_code=400, detail=f"Неизвестный тип: {kind}")

    content = await file.read()

    def _save():
        ref_dir = references.references_dir()
        dest = os.path.join(ref_dir, target_name)
        with open(dest, "wb") as out:
            out.write(content)
        return references.load_from_disk()

    try:
        return await run_blocking(_save)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
